### ISSUE TITLE
Stellantis Peugeot e-208 / Opel Corsa

### ISSUE THREAD

**Issue by deadrabbit87** [meatpiHQ/wican-fw#279]
### Discussed in https://github.com/meatpiHQ/wican-fw/discussions/278

<div type='discussions-op-text'>

<sup>Originally posted by **deadrabbit87** December  7, 2024</sup>
Hi, 

i wan't to get rid of the original Stellantis / My-Peugeot App and use WiCan instead of that. 

As i see, theres no template for this vehicle. What i found is [this](https://www.goingelectric.de/forum/viewtopic.php?p=1514148&sid=038f7b3d9279c69e714d3d0d47a95c73#p1514148) thread in which some IDs are mentioned: 
```
#######
####### Setup commands
#######
ATWS ---> Warm reset set back to defaults
ATI ---> Show interface ID
ATE0 ---> Turn off terminal echo
AT@1 --->
ATSP6 ---> Set to protocol 6: ISO 15765-4 CAN (11 bit ID 500 kbaud)
ATAT1 --->
ATL0 ---> Disable line feed
ATS0 ---> Disable sending spaces between hex characters
ATH1 ---> Turn on odb2 header
ATCAF1 ---> Turn on CAN auto formatting
ATSH79B ---> Set request message header 79B
ATFCSH79B ---> Set flow control header (CAN ID)
ATFCSD300000 ---> Set flow control data
ATFCSM1 ---> Set flow control mode; 1 = User defined CAN ID and data

#######
####### Data commands
#######
ATSH6A2 ---> Set the header of transmitted OBD messages to header 6A2
ATFCSH6A2 ---> Set flow control header (CAN ID)
ATCRA682 ---> Set the CAN hardware filter to 682
22D4341 ---> Ambient temperature
22D8EF1 ---> Battery temperature
22D4021 ---> Speed
ATSH6B4 ---> Set the header of transmitted OBD messages to header 6B4
ATFCSH6B4 ---> Set flow control header (CAN ID)
ATCRA694 ---> Set the CAN hardware filter to 694
22D8161 ---> Battery current
22D8151 ---> Battery voltage
22D86F1 ---> Battery minimum voltage
22D8701 ---> Battery maximum voltage
22D4101 ---> SOC calibrated
22D8651 ---> kWh available
22D8601 ---> SOH
22D8101 ---> SOC
```

Is there a way to create a template for this vehicle with this data? 

</div>

**Comment by atze09** [comment]
I already put this into this topic https://github.com/meatpiHQ/wican-fw/issues/119
Haven't had the time to create a template for my corsa (currently fine with soc only).

What need's to be done is read out each value and define the calculation.
So if you read out the CAN and compare with carscanner value the calculation method can be defined.
@meatpiHQ will also support.

**Comment by deadrabbit87** [comment]
I've already done this, the log from CarScanner is attached. I will check, if i can figure out the conversion factors.
[log_soc_5479.txt](https://github.com/user-attachments/files/18058317/log_soc_5479.txt)

**Comment by atze09** [comment]
> I've already done this, the log from CarScanner is attached. I will check, if i can figure out the conversion factors. [log_soc_5479.txt](https://github.com/user-attachments/files/18058317/log_soc_5479.txt)

Great, yes now the conversion factors need to be figured out and put into the expression of the car template

**Comment by atze09** [comment]
`{
	"car_model": "Opel: Corsa",
	"init": "ATE0;ATH1;ATSPA6;ATS0;ATM0;ATAT1;ATSH6B4;ATFCSD300000;ATFCSM1;",
	"pids": [
		{
			"pid": "22D8101",
			"parameters": [
				{
					"name": "SoC",
					"pid_init": "ATCRA694;",
					"expression": "(B3*20)/51",   ??
					"unit": "%",
					"class": "battery"
				}
			]
		},
		{
			"pid": "??",
			"parameters": [
				{
					"name": "Odometer",
					"pid_init": "??;",
					"expression": "([B3:B6])/10",   ??
					"unit": "km",
					"class": "distance"
				}
			]
		},
		{
			"pid":"??",
			"parameters": [
				{
					"name": "??",
					"pid_init": "??;",
					"expression": "(256*B3)+B4",   ??
					"unit": "km",
					"class": "distance"
				}
			]
		}
	]
}`
@meatpiHQ Could you please double check the init part?

**Comment by deadrabbit87** [comment]
The factors are not the main problem. I'm not very deep into CAN Bus, so i could not figure out, which init parameters are necessary. 

This is what i endet up with: 
```json
{
    "car_model": "Peugeot: E-208",
    "init": "ATSP6;ATE0;ATCAF;ATL0",
    "pids": [
        {
            "pid": "22D4101",
            "pid_init":"ATSH6B4",
            "parameters": [
                {
                    "name": "SOC",
                    "expression": "(B6 + B7)/512",
                    "unit": "%",
                    "class": "battery"
                }
            ]
        },
    ]
}
```

As i suspect `B*` stands for the Byte? The answer from the car is for a soc of `54,79%`:
```
>22D4101
6940562D4106D9D
```
`6D9D` in decimal is `28061`. `28061/512=54,81`

**Comment by atze09** [comment]
> The factors are not the main problem. I'm not very deep into CAN Bus, so i could not figure out, which init parameters are necessary.
> 
> This is what i endet up with:
> 
> ```json
> {
>     "car_model": "Peugeot: E-208",
>     "init": "ATSP6;ATE0;ATCAF;ATL0",
>     "pids": [
>         {
>             "pid": "22D4101",
>             "pid_init":"ATSH6B4",
>             "parameters": [
>                 {
>                     "name": "SOC",
>                     "expression": "(B6 + B7)/512",
>                     "unit": "%",
>                     "class": "battery"
>                 }
>             ]
>         },
>     ]
> }
> ```
> 
> As i suspect `B*` stands for the Byte? The answer from the car is for a soc of `54,79%`:
> 
> ```
> >22D4101
> 6940562D4106D9D
> ```
> 
> `6D9D` in decimal is `28061`. `28061/512=54,81%`.
> 
> I'm not sure if this is correct.
> 
> Is it possible in carscanner, to trigger only one Paremeter? Therefore the reverse enginnering would be much easier.

Yes the reverse engineering is possible, if you click only on sensors and the special sensor in the carscanner app you get the response of this single sensor.
If you are in the progress of reverse engineering, battery information (voltage soh and energy) would be very interesting.

**Comment by meatpiHQ** [comment]
answered in https://github.com/meatpiHQ/wican-fw/discussions/278

**Comment by pimvullers** [comment]
Using the e-208 profile for our Opel Corsa-e.
Have a custom filter for SOC while charging which works great.
Only current issue is the odometer.
In the profile it is just two bytes, which would limit it to 65535km, which seems quite limited. Our car is currently closed to 60000.
Additionally, i have cases where 65535 is reported, which seems to me actually an error, since when interpreted as signed 16 bits it would be -1.
So:
1. I believe the data should be interpreted signed, such that the -1 can be filtered out bij having minimum value 0.
2. I guess there should be additional bytes of data. Haven't had time yet to get a trace to see if there are in this frame.

Anyone else already looked into this?

**Comment by deadrabbit87** [comment]
Hi, 

you are right, in case of a mileage above 65.535 km the value couldn't be displayed. 

Maybe you could extend the expression to `B4*65536+B5*256+B6`. 

I didn't check this.
